### ISSUE TITLE
execute_k8s_job fails when native sidecars are used

### ISSUE THREAD

**Issue by OrenLederman** [dagster-io/dagster#23497]
### Dagster version

1.7.16

### What's the issue?

Since Dagster 1.6.x, the `execute_k8s_job` method can fail when the new [k8s native sidecars](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) are used. These sidecars are defined as init containers.

(In our case, istio sets up two init containers: istio-init, a regular init container, and istio-proxy, a native sidecar)

Based on my investigation, the `wait_for_pod` [method](https://github.com/dagster-io/dagster/blob/c87906eeece657f36bdd6dc539365019220b3f90/python_modules/libraries/dagster-k8s/dagster_k8s/client.py#L526) returns before the main container is ready, leading to a API error when Dagster attempts to read the logs. Below you can see the logs based on the examples I created to create this issue:

```
2024-08-07 17:43:21 -0700 - dagster - INFO - spin_k8s_job - 3c0acbe0-06b6-40a3-8052-82a6181bfc8e - run_k8s_job_op - Creating Kubernetes job 982d5b6b8a69adc0460e8aea4e4b6c8a in namespace default...
2024-08-07 17:43:21 -0700 - dagster - INFO - spin_k8s_job - 3c0acbe0-06b6-40a3-8052-82a6181bfc8e - run_k8s_job_op - Waiting for Kubernetes job to finish...
INFO:root:Waiting for pod "982d5b6b8a69adc0460e8aea4e4b6c8a-d9r7m"
INFO:root:Waiting for pod "982d5b6b8a69adc0460e8aea4e4b6c8a-d9r7m" to initialize...
INFO:root:Pod "982d5b6b8a69adc0460e8aea4e4b6c8a-d9r7m" is ready, done waiting
2024-08-07 17:43:32 -0700 - dagster - INFO - spin_k8s_job - 3c0acbe0-06b6-40a3-8052-82a6181bfc8e - run_k8s_job_op - Deleting Kubernetes job 982d5b6b8a69adc0460e8aea4e4b6c8a in namespace default due to exception
2024-08-07 17:43:33 -0700 - dagster - ERROR - spin_k8s_job - 3c0acbe0-06b6-40a3-8052-82a6181bfc8e - 14474 - run_k8s_job_op - STEP_FAILURE - Execution of step "run_k8s_job_op" failed.

dagster._core.errors.DagsterExecutionStepExecutionError: Error occurred while executing op "run_k8s_job_op"::

dagster_k8s.client.DagsterK8sUnrecoverableAPIError: Unexpected error encountered in Kubernetes API Client.

Stack Trace:
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/dagster/_core/execution/plan/utils.py", line 54, in op_execution_error_boundary
    yield
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/dagster/_utils/__init__.py", line 473, in iterate_with_context
    next_output = next(iterator)
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/dagster/_core/execution/plan/compute_generator.py", line 141, in _coerce_op_compute_fn_to_iterator
    result = invoke_compute_fn(
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/dagster/_core/execution/plan/compute_generator.py", line 129, in invoke_compute_fn
    return fn(context, **args_to_pass) if context_arg_provided else fn(**args_to_pass)
  File "/Users/oren/code/dagster-test-k8s/repository.py", line 34, in run_k8s_job_op
    result = execute_k8s_job(
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/dagster/_core/decorator_utils.py", line 203, in wrapped_with_pre_call_fn
    return fn(*args, **kwargs)
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/dagster_k8s/ops/k8s_job_op.py", line 416, in execute_k8s_job
    raise e
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/dagster_k8s/ops/k8s_job_op.py", line 382, in execute_k8s_job
    log_entry = k8s_api_retry(
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/dagster_k8s/client.py", line 128, in k8s_api_retry
    raise DagsterK8sUnrecoverableAPIError(

The above exception was caused by the following exception:
kubernetes.client.exceptions.ApiException: (400)
Reason: Bad Request
HTTP response headers: HTTPHeaderDict({'Audit-Id': 'a1307c9f-3d8d-4f13-91ce-08b89c2705c1', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Date': 'Thu, 08 Aug 2024 00:43:32 GMT', 'Content-Length': '223'})
HTTP response body: b'{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"container \\"dagster\\" in pod \\"982d5b6b8a69adc0460e8aea4e4b6c8a-d9r7m\\" is waiting to start: PodInitializing","reason":"BadRequest","code":400}\n'


Stack Trace:
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/dagster_k8s/client.py", line 109, in k8s_api_retry
    return fn()
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/dagster_k8s/ops/k8s_job_op.py", line 383, in <lambda>
    lambda: next(log_stream),
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/kubernetes/watch/watch.py", line 176, in stream
    resp = func(*args, **kwargs)
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/kubernetes/client/api/core_v1_api.py", line 23957, in read_namespaced_pod_log
    return self.read_namespaced_pod_log_with_http_info(name, namespace, **kwargs)  # noqa: E501
  File "/Users/oren/Library/Caches/pypoetry/virtualenvs/dagster-test-k8s-KKSQEs2L-py3.10/lib/python3.10/site-packages/kubernetes/client/api/core_v1_api.py", line 24076, in read_namespaced_pod_log_with_http_info
    return self.api_client.call_api(
```

From what I can tell, the current logic in `wait_for_pod` iterates over all the containers. It's expecting to see all the init containers [successfully complete](https://github.com/dagster-io/dagster/blob/c87906eeece657f36bdd6dc539365019220b3f90/python_modules/libraries/dagster-k8s/dagster_k8s/client.py#L683). Then, it's looking for the first container to show a [ready state](https://github.com/dagster-io/dagster/blob/c87906eeece657f36bdd6dc539365019220b3f90/python_modules/libraries/dagster-k8s/dagster_k8s/client.py#L615) before it determines that the pod is ready.
When a native sidecar is used, however, this logic breaks. Because the code doesn't distinguish between init containers and real containers, it will see a sidecar (an init container) with a ready status and decide that the pod is ready.

### What did you expect to happen?

I expect `wait_for_pod` to wait for all init containers to be done (exited successfully for regular init containers) or ready (for native sidecars), and only then check the status of the main dagster container.

### How to reproduce?

Here's a minimalistic code that recreates the issue. In this example, the first init container is a native side car (`istio-proxy-sidecar`). This container starts successfully and is "ready". The second init container (which is a regular init container) hasn't completed yet, but `wait_for_pod` returns anyway, leading to the failure I described.

```
from dagster import (
    job,
    op,
    Definitions,
    get_dagster_logger,
)
from dagster_k8s import execute_k8s_job
import logging

logger = logging.getLogger()
logger.setLevel("INFO")

@op()
def run_k8s_job_op(context):
    job_name = "wait-job"
    

    # Define the init container
    init_containers = [
        {
        "name": "istio-proxy-sidecar",
        "image": "busybox",
        "command": ["sh", "-c", "echo 'Sidecar container running' && tail -f /dev/null"],
        "restartPolicy": "Always" # that's how you define a native sidecar
        },
        {
        "name": "regular-init-container",
        "image": "busybox",
        "command": ["sh", "-c", "echo 'Init container running' && sleep 600"]
        }
    ]

    # Execute a k8s job that just waits for 15 minutes
    result = execute_k8s_job(
        context,
        image="busybox",
        args=["sh", "-c", "sleep 900"],
        load_incluster_config=False,
        namespace="default",
        pod_spec_config={
            "initContainers": init_containers,
        },
    )
    
    return result

@job
def spin_k8s_job():
    run_k8s_job_op()

defs = Definitions(
    jobs=[spin_k8s_job],
)
```

### Deployment type

Dagster Helm chart

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by OrenLederman** [comment]
As a temporary workaround, monkey-patching `wait_for_pod` so it set my init containers as `ignore_containers` seems to be working:

```
from dagster_k8s.client import DagsterKubernetesClient

original_wait_for_pod = DagsterKubernetesClient.wait_for_pod

def patched_wait_for_pod(*args, **kwargs):
    kwargs['ignore_containers'] = {"istio-init", "istio-proxy"}
    return original_wait_for_pod(*args, **kwargs)

DagsterKubernetesClient.wait_for_pod = patched_wait_for_pod
```

**Comment by michaelromagne** [comment]
Hi @OrenLederman , is this issue priorized ?

I have the same issue and the fix you propose work for the toy example above but not for deployments with Helm

**Comment by garethbrickman** [comment]
@michaelromagne It is not being actively worked on at this time, but does the solution in your other issue work in its place? https://github.com/dagster-io/dagster/issues/24256

**Comment by easontm** [comment]
@michaelromagne @garethbrickman I have submitted a PR for this one :)

**Comment by michaelromagne** [comment]
Thanks @easontm that's great :partying_face: 

@garethbrickman will this fix be released with 1.8.8 ? When is 1.8.8 planned ? Thanks :smiley:

**Comment by garethbrickman** [comment]
@michaelromagne I believe so. 1.8.8 will be released later today.

**Comment by OrenLederman** [comment]
Re-opening this bug because https://github.com/dagster-io/dagster/pull/24432 does not fix the specific issue with the new sidecar init containers. Instead, it breaks the monkey-patch workaround:
* `initcontainers` doesn't filter out the containers in `ignore_containers`
* later, when the main `dagster` container is read, the code gets to `if initcontainers.issubset(exited_containers | ready_containers)` and evaluates is as `false` (since the init containers I ignore are not in the exited_containers or ready_containers list). Then the code does another iteration in the loop and crashes because `container_status = next(s for s in all_statuses if s.name not in exited_containers | ready_containers)` runs on an empty list and raises a StopIteration exception

So basically, there are two bugs now:
* sidecar init containers that reach `ready` state will still cause the loop to exist prematurely (it confuses them for the main dagster container)
* ignore_containers are't properly ignored (for init containers) . This seems to be an issue in general, but it also breaks the workaround that I found

I'll try to come up with an easy way to re-create the issue after the recent changes, and see if I can add a fix and/or unit tests. The real solution might be to handle the init containers in a separate loop, and only check the status of the main container after we make sure they are all ready (for sidecars) and exited (for regular init containers).

**Comment by OrenLederman** [comment]
I was unable to create a minimal example showing where the existing code fails, but when debugging the code I noticed that the code doesn't really wait for the main container to become ready. The existing unit test for init containers were incomplete, so they did not capture this issue. I published a PR that should fix the issue.

**Comment by OrenLederman** [comment]
In the meantime, if anyone is experiencing a similar issue, you can monkey-patch the problematic code using the following code:
```
from typing import Any, Optional, Set
from dagster_k8s.client import (
    DagsterKubernetesClient,
    WaitForPodState,
    DEFAULT_WAIT_TIMEOUT,
    DEFAULT_WAIT_BETWEEN_ATTEMPTS,
    DagsterK8sError,
    KubernetesWaitingReasons,
)
from dagster import (
    _check as check,
)

# Monkey patch the wait_for_pod as a workaround to the issue we described here - https://github.com/dagster-io/dagster/issues/23497
original_wait_for_pod = DagsterKubernetesClient.wait_for_pod


# This code is based on the original wait_for_pod method here (version 1.9.3) -https://github.com/dagster-io/dagster/blob/12b54a13359dd2691c7a0c68d758a3fa321d0a3a/python_modules/libraries/dagster-k8s/dagster_k8s/client.py#L567
# The only change here is to continue the loop if the "ready" container is an init container, and wait for the main container to be ready.
def patched_wait_for_pod(
    self,
    pod_name: str,
    namespace: str,
    wait_for_state: WaitForPodState = WaitForPodState.Ready,
    pod_launch_timeout: float = DEFAULT_WAIT_TIMEOUT,
    wait_timeout: float = DEFAULT_WAIT_TIMEOUT,
    wait_time_between_attempts: float = DEFAULT_WAIT_BETWEEN_ATTEMPTS,
    start_time: Any = None,
    ignore_containers: Optional[Set] = None,
) -> None:
    """Wait for a pod to launch and be running, or wait for termination (useful for job pods).

    Args:
    ----
        pod_name (str): Name of the pod to wait for.
        namespace (str): Namespace in which the pod is located.
        wait_for_state (WaitForPodState, optional): Whether to wait for pod readiness or
            termination. Defaults to waiting for readiness.
        pod_launch_timeout (numeric, optional): Timeout after which to give up and raise exception
            if the pod never appears. Defaults to DEFAULT_WAIT_TIMEOUT. Set to 0 to disable.
        wait_timeout (numeric, optional): Timeout after which to give up and raise exception.
            Defaults to DEFAULT_WAIT_TIMEOUT. Set to 0 to disable.
        wait_time_between_attempts (numeric, optional): Wait time between polling attempts. Defaults
            to DEFAULT_WAIT_BETWEEN_ATTEMPTS.
        start_time (numeric, optional): The start time of the wait, used for testing.
        ignore_containers (set, optional): The container names that we should ignore
            when waiting for the pod to be ready/terminate.

    Raises:
    ------
        DagsterK8sError: Raised when wait_timeout is exceeded or an error is encountered

    """
    check.str_param(pod_name, "pod_name")
    check.str_param(namespace, "namespace")
    check.inst_param(wait_for_state, "wait_for_state", WaitForPodState)
    check.numeric_param(wait_timeout, "wait_timeout")
    check.numeric_param(pod_launch_timeout, "pod_launch_timeout")
    check.numeric_param(wait_time_between_attempts, "wait_time_between_attempts")

    self.logger(f'Waiting for pod "{pod_name}"')

    start = start_time or self.timer()

    # A set of container names that have exited.
    exited_containers = set()
    ready_containers = set()
    ignore_containers = ignore_containers or set()
    error_logs = []

    while True:
        pods = self.core_api.list_namespaced_pod(
            namespace=namespace, field_selector=f"metadata.name={pod_name}"
        ).items
        pod = pods[0] if pods else None

        if pod_launch_timeout and self.timer() - start > pod_launch_timeout:
            raise DagsterK8sError(
                f"Timed out while waiting for pod to become ready with pod info: {pod!s}"
            )

        if pod is None:
            self.logger(f'Waiting for pod "{pod_name}" to launch...')
            self.sleeper(wait_time_between_attempts)
            continue

        if not pod.status.init_container_statuses and not pod.status.container_statuses:
            self.logger(
                "Waiting for pod init_container or container status to be set by kubernetes..."
            )
            self.sleeper(wait_time_between_attempts)
            continue
        break

    while True:
        pods = self.core_api.list_namespaced_pod(
            namespace=namespace, field_selector=f"metadata.name={pod_name}"
        ).items
        pod = pods[0] if pods else None
        if pod is None:
            raise DagsterK8sError(f'Pod "{pod_name}" was unexpectedly killed')

        if wait_timeout and self.timer() - start > wait_timeout:
            raise DagsterK8sError(
                f"Timed out while waiting for pod to get to status {wait_for_state.value} with pod info: {pod!s}"
            )
        # https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#containerstatus-v1-core
        all_statuses = []
        all_statuses.extend(pod.status.init_container_statuses or [])
        all_statuses.extend(pod.status.container_statuses or [])
        initcontainers = set(s.name for s in (pod.status.init_container_statuses or []))

        # Filter out ignored containers
        all_statuses = [s for s in all_statuses if s.name not in ignore_containers]

        # Always get the first status from the list, which will first get the
        # init container (if it exists), then will iterate through the loop
        # of all containers if we are waiting for termination.
        #
        # In case we are waiting for the pod to be ready, we will exit after
        # the first container in this list is ready.
        container_status = next(
            s for s in all_statuses if s.name not in exited_containers | ready_containers
        )

        # State checks below, see:
        # https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#containerstate-v1-core
        state = container_status.state

        if state.running is not None:
            if wait_for_state == WaitForPodState.Ready:
                # ready is boolean field of container status
                ready = container_status.ready
                if not ready:
                    self.logger(f'Waiting for pod "{pod_name}" to become ready...')
                    self.sleeper(wait_time_between_attempts)
                    continue
                else:
                    ready_containers.add(container_status.name)

                    # If it's a ready *sidecar* init container, we need to continue and wait for the main container to be ready
                    if container_status.name in initcontainers:
                        self.logger(
                            f'Init container "{container_status.name}" is ready, waiting for the main container to be ready...'
                        )
                        continue

                    if initcontainers.issubset(exited_containers | ready_containers):
                        self.logger(f'Pod "{pod_name}" is ready, done waiting')
                        break

            else:
                check.invariant(
                    wait_for_state == WaitForPodState.Terminated, "New invalid WaitForPodState"
                )
                self.sleeper(wait_time_between_attempts)
                continue

        elif state.waiting is not None:
            # https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#containerstatewaiting-v1-core
            if state.waiting.reason == KubernetesWaitingReasons.PodInitializing:
                self.logger(f'Waiting for pod "{pod_name}" to initialize...')
                self.sleeper(wait_time_between_attempts)
                continue
            if state.waiting.reason == KubernetesWaitingReasons.CreateContainerConfigError:
                self.logger(
                    f'Pod "{pod_name}" is waiting due to a CreateContainerConfigError with message "{state.waiting.message}"'
                    " - trying again to see if it recovers"
                )
                self.sleeper(wait_time_between_attempts)
                continue
            elif state.waiting.reason == KubernetesWaitingReasons.ContainerCreating:
                self.logger("Waiting for container creation...")
                self.sleeper(wait_time_between_attempts)
                continue
            elif state.waiting.reason is None:
                self.logger(
                    f'Pod "{pod_name}" is waiting with reason "None" - this is temporary/transition state'
                )
                self.sleeper(wait_time_between_attempts)
                continue
            elif state.waiting.reason in [
                KubernetesWaitingReasons.ErrImagePull,
                KubernetesWaitingReasons.ImagePullBackOff,
                KubernetesWaitingReasons.CrashLoopBackOff,
                KubernetesWaitingReasons.RunContainerError,
            ]:
                debug_info = self.get_pod_debug_info(pod_name, namespace, pod=pod)
                raise DagsterK8sError(
                    f'Failed: Reason="{state.waiting.reason}"'
                    f' Message="{state.waiting.message}"\n{debug_info}'
                )
            else:
                raise DagsterK8sError(f"Unknown issue: {state.waiting}")

        # https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#containerstateterminated-v1-core
        elif state.terminated is not None:
            container_name = container_status.name
            if state.terminated.exit_code != 0:
                raw_logs = self.retrieve_pod_logs(
                    pod_name, namespace, container_name=container_name
                )
                message = state.terminated.message
                msg = (
                    f'Container "{container_name}" failed with message: "{message}" '
                    f'and pod logs: "{raw_logs}"'
                )

                self.logger(msg)
                error_logs.append(msg)
            elif container_name in initcontainers:
                self.logger(
                    f"Init container {container_name} in {pod_name} has exited successfully"
                )
            else:
                self.logger(f"Container {container_name} in {pod_name} has exited successfully")

            exited_containers.add(container_name)
            if len(all_statuses) != len(exited_containers):
                continue

            if error_logs:
                logs = "\n\n".join(error_logs)
                raise DagsterK8sError(
                    f"Pod {pod_name} terminated but some containers exited with errors:\n{logs}"
                )
            else:
                self.logger(f"Pod {pod_name} exited successfully")
            break

        else:
            raise DagsterK8sError("Should not get here, unknown pod state")


DagsterKubernetesClient.wait_for_pod = patched_wait_for_pod
```
